### PR TITLE
[Feature] MXFP4 model loading

### PR DESCRIPTION
**Pull request by fxmarty-amd**
This PR allows to load MXFP4 models in sglang, using dynamic per-group OCP MXFP4 quantization for activations for linear layers.

The supported models are quantized using AMD Quark.

For now, the execution of GEMM is simulated on fp16/bf16, but in the future mxfp4 kernels will be added.

I get sensibly similar eval results vs the recent integration in vllm, on wikitext for a llama 2 70B mxfp4 model:

sglang (this PR):
| Tasks  |Version|Filter|n-shot|    Metric     |   |Value |   |Stderr|
|--------|------:|------|-----:|---------------|---|-----:|---|------|
|wikitext|      2|none  |     0|bits_per_byte  |↓  |0.5560|±  |   N/A|
|        |       |none  |     0|byte_perplexity|↓  |1.4701|±  |   N/A|
|        |       |none  |     0|word_perplexity|↓  |7.8516|±  |   N/A|

vllm:
| Tasks  |Version|Filter|n-shot|    Metric     |   |Value |   |Stderr|
|--------|------:|------|-----:|---------------|---|-----:|---|------|
|wikitext|      2|none  |     0|bits_per_byte  |↓  |0.5565|±  |   N/A|
|        |       |none  |     0|byte_perplexity|↓  |1.4707|±  |   N/A|
|        |       |none  |     0|word_perplexity|↓  |7.8674|±  |   N/A|

To do:
- [x] Documentation
- [x] Test
- [x] Support MOE models (likely in an other PR)
- [ ] Support mxfp4 * mxfp4 GEMM kernel (likely in an other PR)
- [x] Test eval parity for Deepseek R1, Llama 4 and llama 405B with vllm

### CODE REVIEW

**Comment by HaiShaw** [comment]
Can we extend this to handle fp8/bf16/fp16 activations (with only weight be mxfp4)?

**Comment by zhaochenyang20** [comment]
@fxmarty-amd rebase？

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -1463,6 +1464,11 @@ def __init__(
         prefix: str = "",
     ) -> None:
         super().__init__()
+
+        fuse_qkv_a_proj = hasattr(config, "q_lora_rank") and config.q_lora_rank is not None
+        if fuse_qkv_a_proj:
+            self.packed_modules_mapping["fused_qkv_a_proj_with_mqa"] = ["q_a_proj", "kv_a_proj_with_mqa"]
+
```
Comment (BowenBao) [inline python/sglang/srt/models/deepseek_v2.py]: are these additions needed solely by quark?

File: python/sglang/srt/model_loader/weight_utils.py [hunk 2]
```diff
@@ -728,6 +729,16 @@ def maybe_remap_kv_scale_name(name: str, params_dict: dict) -> Optional[str]:
...
+    
+    quark_scale_names = {
+        ".q_proj.output_scale": ".attn.q_scale",
+        ".k_proj.output_scale": ".attn.k_scale",
+        ".v_proj.output_scale": ".attn.v_scale",
+        "self_attn.prob_output_scale": ".attn.prob_scale"
+    }
+    for quark_scale_name, sglang_scale_name in quark_scale_names.items():
+        if name.endswith(quark_scale_name):
+            return name.replace(quark_scale_name, sglang_scale_name)
```
Comment (BowenBao) [inline python/sglang/srt/model_loader/weight_utils.py]: is this how is done in vllm?

**Comment by BowenBao** [comment]
> Can we extend this to handle fp8/bf16/fp16 activations (with only weight be mxfp4)?

These are some good ideas, however I'd suggest focus the scope of this PR within just mxfp4 weight and dynamic mxfp4 activations. Folks can fill in these two spots with mxfp4 kernel call to make it work e2e.

https://github.com/sgl-project/sglang/blob/d21830c76d51313693b67738bdb19b41eed2c856/python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py#L106

https://github.com/sgl-project/sglang/blob/d21830c76d51313693b67738bdb19b41eed2c856/python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py#L815

File: python/sglang/srt/model_loader/weight_utils.py (hunk 2 above)
Comment (fxmarty-amd) [reply python/sglang/srt/model_loader/weight_utils.py]: No, sglang has a different logic.

vllm uses https://github.com/vllm-project/vllm/blob/d637b960994119907b41c82d79f5a71c96dd419b/vllm/model_executor/layers/quantization/quark/quark.py#L344, which does not exist for other quantization schemes in sglang.

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (fxmarty-amd) [reply python/sglang/srt/models/deepseek_v2.py]: This is needed for https://github.com/fxmarty-amd/sglang/blob/c0e1e1eeb12bf69ca28953d1c0960f370e7ee645/python/sglang/srt/layers/quantization/quark/quark.py#L64-L66

and https://github.com/sgl-project/sglang/blob/844e2f227ab0cce6ef818a719170ce37b9eb1e1b/python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors.py#L124-L126 to properly work.

**Comment by fxmarty-amd** [comment]
@zhaochenyang20 I merged main, let me know what you think!

**Comment by fxmarty-amd** [comment]
Hi @zhaochenyang20, happy to update if you have comments

**Comment by xutizhou** [comment]
Hi @fxmarty-amd, thank you very much for your outstanding work! I was wondering, is Quark also compatible with NVIDIA GPUs?

File: python/sglang/srt/configs/model_config.py [hunk 3]
```diff
@@ -327,6 +327,7 @@ def _verify_quantization(self) -> None:
             "compressed-tensors",
             "fbgemm_fp8",
             "w8a8_fp8",
+            "quark",
```
Comment (HaiShaw) [inline python/sglang/srt/configs/model_config.py:395]: please consider to change `quark` to something more specific, e.g. `quark_mxfp4a16`

File: python/sglang/srt/layers/quantization/mxfp4_utils.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+OCP_MX_BLOCK_SIZE = 32
+
+
+def dequant_mxfp4(x: torch.Tensor, scale: torch.Tensor, float_dtype: torch.dtype) -> torch.Tensor:
+    try:
+        from quark.torch.kernel import mx
+    except ImportError as err:
+        raise ImportError("The package `amd-quark` is required to use "
+                        "MX-FP4 models. Please install it with `pip install "
+                        "amd-quark`.") from err
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/mxfp4_utils.py]: Pls modify `python/pyproject.toml` to have `amd-quark` (with a versioned one) installed to hip case.

File: python/sglang/srt/layers/quantization/mxfp4_utils.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# SPDX-License-Identifier: Apache-2.0
+import torch
+
+OCP_MX_BLOCK_SIZE = 32
+
+
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/mxfp4_utils.py:6]: add input/usage comment lines as this api header

File: python/sglang/srt/layers/quantization/mxfp4_utils.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        from quark.torch.kernel import mx
+    except ImportError as err:
+        raise ImportError("The package `amd-quark` is required to use "
+                        "MX-FP4 models. Please install it with `pip install "
+                        "amd-quark`.") from err
+
+    return mx.dq_mxfp4(x, scale, float_dtype)
+
+
+def quant_dequant_mxfp4(x: torch.Tensor, scale_calculation_mode: str = "even") -> torch.Tensor:
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/mxfp4_utils.py]: add input/usage comment lines (e.g. what modes, and what is meaning) as this api header

File: python/sglang/srt/layers/quantization/quark/quark.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/quark.py:1]: This file and classes need to be specific (not quark), `quark_` can be prefix, appropriate example can be `quark_mxfp4_`

File: python/sglang/srt/layers/quantization/quark/quark_moe.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/quark_moe.py:1]: same pls be specific (to name and namespace), quark can be a lot more than mxfp4

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py:1]: Design logic should be (with mxfp4 weight):
- On HW does not support mxfp matmul, we need to use matmul (and activation dtype) of fp16/bf16 or fp8, instructed by `SGLANG_QUARK_MXFP_DQ16` or `SGLANG_QUARK_MXFP_DQ8`, remove the unclear name `SGLANG_QUARK_EMU_MEM_OPT`
- On HW support mxfp matmal, quantize activation to mxfp, and call mxfp matmul (this is separate from this PR)

**Review by HaiShaw — CHANGES_REQUESTED** [review]
1. Please resolve conflicts
2. Internal choice of w4a4_mxfp4 on HW without mxfp support needs to be revised (see comment inside). The only good assumption is that weight is given in mxfp, from there if we see that HW is legacy, then we only do a16 or a8, but not a4.

File: python/sglang/srt/utils.py [hunk 10]
```diff
@@ -1489,6 +1490,15 @@ def get_compiler_backend() -> str:
 def supports_custom_op() -> bool:
     return hasattr(torch.library, "custom_op")
 
+def supports_mx() -> bool:
```
Comment (HaiShaw) [inline python/sglang/srt/utils.py:1753]: change this to `mxfp_supported()`

**Review by HaiShaw — CHANGES_REQUESTED** [review]

File: python/sglang/srt/configs/model_config.py (hunk 3 above)
Comment (BowenBao) [reply python/sglang/srt/configs/model_config.py:395]: Hey @HaiShaw, `quark` format is designed to be flexible, the actual quantized layers and configurations are defined in the model's config.json, and they can be mixed.

File: python/sglang/srt/layers/quantization/quark/quark.py (hunk 7 above)
Comment (BowenBao) [reply python/sglang/srt/layers/quantization/quark/quark.py:1]: same as earlier reply, here for each layer we will dispatch to the matching quantized layer based on the info in config. For individual quant spec implementations, they reside under the scheme folder and they are named in the style like `quark_mxfp4_` you mentioned.

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py (hunk 9 above)
Comment (BowenBao) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py:1]: Yea quark would not dictate the kernel to be used. The current implementation allows end-to-end runs and validation for mxfp4-mxfp4 gemm on HW that does not support it. The flag is controlling a trade-off between better perf and lower memory usage. The flag can be dropped if you feel it's not making sense.

fp16/bf16 or fp8 kernels can be follow-ups but I'm not sure how important they are, unless there are efficient fp8/fp16-mxfp4 kernels and a use case for them?

**Comment by HaiShaw** [comment]
@fxmarty-amd let's rebase

File: python/sglang/srt/utils.py (hunk 10 above)
Comment (fxmarty-amd) [reply python/sglang/srt/utils.py:1753]: why? there are `supports_custom_op`, `support_triton` above

File: python/sglang/srt/configs/model_config.py (hunk 3 above)
Comment (fxmarty-amd) [reply python/sglang/srt/configs/model_config.py:395]: See e.g. https://huggingface.co/amd/Llama-3.1-405B-Instruct-MXFP4-Preview/blob/main/config.json#L225

We do not make use of `QuantizationConfig.override_quantization_method` in our case.

https://github.com/sgl-project/sglang/blob/4a8837950abb7a39d5b890b8f4ee21bd9ded959d/python/sglang/srt/configs/model_config.py#L423-L436

File: python/sglang/srt/layers/quantization/mxfp4_utils.py (hunk 4 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/mxfp4_utils.py]: done

File: python/sglang/srt/layers/quantization/mxfp4_utils.py (hunk 5 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/mxfp4_utils.py:6]: done

File: python/sglang/srt/layers/quantization/mxfp4_utils.py (hunk 6 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/mxfp4_utils.py]: done

File: python/sglang/srt/layers/quantization/quark/quark.py (hunk 7 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark/quark.py:1]: In spirit this is similar to https://github.com/fxmarty-amd/sglang/tree/main/python/sglang/srt/layers/quantization/compressed_tensors (which might have been inspired from quark in vllm, or the reverse...)

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py (hunk 9 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4.py:1]: @HaiShaw @BowenBao Thank you for your comments. I removed `SGLANG_QUARK_EMU_MEM_OPT` as the similar env variable was removed in vllm as well for simplicity. We'll just dequantize on the fly when running on gfx942 (or gfx950 as long as kernels are not integrated).

> On HW does not support mxfp matmul, we need to use matmul (and activation dtype) of fp16/bf16 or fp8, instructed by SGLANG_QUARK_MXFP_DQ16 or SGLANG_QUARK_MXFP_DQ8, remove the unclear name SGLANG_QUARK_EMU_MEM_OPT

Eventually no additional env variable are added in this PR, after 532697fdf8d67fe92580bef1889b1d29da7a4aa0.

I agree we could "dequantize" to fp8 instead of fp16/bf16, but this is out of scope of this PR for simulated mxfp4 execution.

**Comment by fxmarty-amd** [comment]
Hi @HaiShaw, conflicts should have been solved, and tests from https://github.com/fxmarty-amd/sglang/blob/mxfp4/test/srt/models/test_quark_models.py run fine for me locally.

**Comment by fxmarty-amd** [comment]
@xutizhou

> Hi @fxmarty-amd, thank you very much for your outstanding work! I was wondering, is Quark also compatible with NVIDIA GPUs?

Quark quantizer should run just fine on Nvidia GPUs, although our CI runs on AMD Instinct GPUs.

**Comment by fxmarty-amd** [comment]
Hi @HaiShaw @zhyncs @zhaochenyang20 hope you are doing well.

I solved conflicts again and made sure the added tests pass successfully as well.

Let me know if you need anything from me to get this PR merged.

Edit: actually, TestR1MXFP4Accuracy is not passing now - checking.

**Comment by fxmarty-amd** [comment]
gentle ping @zhyncs @zhaochenyang20 

The added tests pass for me.

I disabled `--attention-backend aiter` for now as there are ongoing issues unrelated to this PR investigated by @kkHuang-amd .

**Comment by BowenBao** [comment]
@HaiShaw please kindly take a look.

**Comment by fxmarty-amd** [comment]
Closing as https://github.com/sgl-project/sglang/pull/8255 will land instead.
